compiler/rustc_const_eval/src/interpret/machine.rs

```diff
@@ -304,6 +304,16 @@ pub trait Machine<'mir, 'tcx: 'mir>: Sized {
         Ok(())
     }
 
+    /// Called before any global allocation is accessed for reading.
+    /// Used by const eval to prevent reading from the static root allocation.
+    #[inline]
+    fn check_alloc_read_access(
```

Comment:
If this is only for globals, the name should reflect that. But from the implementation it looks to be called for everything? Why is before_memory_read not sufficient?

Reply:
As you noticed in https://github.com/rust-lang/rust/pull/116564#discussion_r1361629164, it is not called everywhere, Most notably, it is not called for ZSTs. I think I found a fix for that, so re-using `before_memory_read` by adding a `kind` argument to it